lib/Mojolicious/Plugin/OpenAPI.pm

```diff
@@ -144,28 +140,26 @@ sub _add_routes {
 
 sub _before_render {
   my ($c, $args) = @_;
-  return unless $args->{template};    # This and status is set in the case of 404 and 500
-  return unless my $status = $args->{status} || $c->stash('status');
-  return unless my $self = _self($c);
-
-  if ($status eq '500') {
-    $args->{data} = $self->_renderer->(
-      $c, {errors => [{message => 'Internal server error.', path => '/'}], status => 500}
-    );
-    $args->{status} = $c->stash('status') || $status;
-  }
-  elsif (!$c->stash('openapi.op_path')) {
-    $status = 404;
-    $args->{data}
-      = $self->_renderer->($c, {errors => [{message => 'Not found.', path => '/'}], status => 404});
-    $args->{status} = $c->stash('status') || $status;
-  }
-  elsif ($status eq '404') {
-    $status = 501;
-    $args->{data} = $self->_renderer->(
-      $c, {errors => [{message => 'Not implemented.', path => '/'}], status => 501}
+  my $handler = $args->{handler} || 'openapi';
+
+  # Call _render() for response data
+  return if $handler eq 'openapi' and exists $c->stash->{openapi} or exists $args->{openapi};
+
+  # Fallback to default handler for things like render_to_string()
+  return $args->{handler} = $c->app->renderer->default_handler unless exists $args->{handler};
+
+  # Call _render() for errors
+  my $status = $args->{status} || $c->stash('status') || '200';
+  if ($handler eq 'openapi' and ($status eq '404' or $status eq '500')) {
+    $args->{handler} = 'openapi';
+    $args->{status} = ($status eq '404' and $c->stash('openapi.op_path')) ? 501 : $status;
+    $c->stash(
+      status  => $args->{status},
+      openapi => {
+        errors => [{message => $c->res->default_message($args->{status}) . '.', path => '/'}],
+        status => $args->{status},
+      }
```

Comment:
Setting "openapi" in the stash will make sure the internally generated error documents will also be passed through the output validator.